Various edits throughout the specification (#122)
dcdpr/did-btc1: chapters/CRUD-Operations.md

#### From External Intermediate DID Document

The From External ::intermediate DID document:: algorithm enables the ability 
to create a **did:btc1** from an external ::intermediate DID document::. This allows for a
more complex ::initial DID document::, including the ability to include
Service Endpoints and ::Beacons:: that support aggregation.

It takes the following inputs:

* `intermediateDocument` - any arbitrary, valid DID document with the `identifier`
   replaced with the placeholder value throughout all fields (e.g. the id field)
   `did:btc1:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx`. It
   SHOULD include at least one `verificationMethod` and `service` of the type

>    [SingletonBeacon]; REQUIRED; object.

Reviewer: What does the [ ] do?

Author: I think I meant to do ::SingletonBeacon::

#### From External Intermediate DID Document

The From External ::Intermediate DID Document:: algorithm enables the ability 
to create a **did:btc1** from some external intermediate DID document. This allows for a
more complex ::Initial DID Document::, including the ability to include

> Service Endpoints and ::Beacons:: that support aggregation.

Reviewer: ```suggestion
Service Endpoints and ::Beacons:: that support aggregation.
```

@jenniemeier should this be capitalized? Not sure where we landed

It takes the following inputs:

* `intermediateDocument` - any arbitrary, valid DID document with the `identifier`
   replaced with the placeholder value throughout all fields (e.g. the id field)
   `did:btc1:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx`. It

>    should include at least one verificationMethod and service of the type

Reviewer 2: I noticed that this change (to lowercase) made the line consistent with other uses of "should" within the spec. Is the recommendation to capitalize all normative uses, or just this one?

>    should include at least one verificationMethod and service of the type

Author: Yeah, I think I normalize to the wrong case. All should be normalized to CAPS. I'll add that to the PR.

   [SingletonBeacon].
* `version` - the identifier version, optional, defaults to 1.

> * `network` - the bitcoin network where the DID and DID Document can be resolved,

Author: I think the recent move has been to lower case the "d" in "document" throughout the spec. Although I recall @dcdpr questioning this in our last call.

> * `network` - the bitcoin network where the DID and DID Document can be resolved,

Reviewer: No, definitely lower case. We should copy the style that DID core uses here. As we are doing throughout the spec

```suggestion
* `network` - the bitcoin network where the DID and DID document can be resolved,
```

The Read operation is an algorithm consisting of a series of subroutine algorithms
executed by a resolver after a resolution request identifying a specific
**did:btc1** `identifier` is received from a client at ::Resolution Time::.

> The request MUST always contain the `resolutionOptions` object containing additional
> information to be used in resolution. This object MAY be empty. 

Reviewer: ```suggestion
The request MUST always contain the `resolutionOptions` object containing additional
information to be used in resolution. This object MAY be empty. 
See the [DID Resolution specification](https://w3c.github.io/did-resolution/#did-resolution-options) for further details about the DID Resolution Options object.
```

I am sure there are many places throughout the spec that could benefit from better referencing. This seems like an easy one to catch now

The resolver then attempts to resolve the
DID document of the `identifier` at a specific ::Target Time::. The
::Target Time:: is either provided in `resolutionOptions` or is set to the
::Resolution Time:: of the request.

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier

> * `resolutionOptions` - an optional object containing the following options

Reviewer: ```suggestion
* `resolutionOptions` - an object containing the following optional options
```

#### Resolve Initial DID Document

> This algorithm specifies how to resolve an initial DID document and validate

Reviewer: ```suggestion
This algorithm specifies how to resolve an ::initial DID document:: and validate
```

it against the `identifier` for a specific **did:btc1**.

It takes the following inputs:

* `identifier` - A valid **did:btc1** identifier
* `identifierComponents` - An object containing the following properties
  * `idType` - the type of identifier (`KEY` or `EXTERNAL`)
  * `version` - the identifier version

>   * `network` - the identifier bitcoin network

Reviewer: ```suggestion
  * `network` - the identifier Bitcoin network
```

I believe

  * `genesisBytes` - the originating public key bytes
* `resolutionsOptions` - optional object that extends the default
   [DID Resolution Options](https://www.w3.org/TR/did-1.0/#did-resolution-options)
  * `versionId` - TODO: What exactly is this?
  * `versionTime` - the time at which the identifier and DID document were created
  * `sidecarData` - data necessary for resolving a DID such as ::DID Update Payloads:: and ::SMT proofs::.

>   * `network` - the identifier bitcoin network

Reviewer: This is nicer than above. We should use this style both times I feel.

Also this TODO for versionId. versionId in did:btc1 is an incrementing integer starting from 1, the version of the initial document

::Initial DID Document:: from a secp256k1 public key.

It takes the following inputs:

* `identifier` - A valid **did:btc1** identifier
* `identifierComponents` - An object containing the following properties
  * `idType` - the type of identifier (KEY or EXTERNAL)
  * `version` - the identifier version

>   * `network` - the identifier bitcoin network

Reviewer: ```suggestion
  * `network` - the identifier Bitcoin network
```

1. Else MUST raise `invalidHRPValue` error.
1. Return `initialDocument`.

##### Deterministically Generate Initial DID Document

The Deterministically Generate ::initial DID document:: algorithm generates an
::initial DID document:: from a secp256k1 public key.

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier; REQUIRED; string
* `identifierComponents` - The decoded parts of a **did:btc1** identifier; REQUIRED; object
  * `idType` - the type of identifier (`KEY` or `EXTERNAL`); REQUIRED; string
  * `version` - the identifier version; REQUIRED; integer
  * `network` - the Bitcoin network used for the identifier; REQUIRED; string
  * `genesisBytes` - the originating public key; REQUIRED; bytes

It returns the following output:

> * `initialDocument` - the valid first version of a DID document for a given btc1 identifier.

Reviewer: ```suggestion
* `initialDocument` - the valid first version of a DID document for a given *did:btc1*** identifier.
```

The Deterministically Generate Beacon Services algorithm generates three
::Beacons:: from the single `keyBytes` value used to generate the deterministic
**did:btc1**, one for each of the following three Bitcoin address types for the
Bitcoin `network` specified by the DID:

* Pay-to-Public-Key-Hash (P2PKH)
* Pay-to-Witness-Public-Key-Hash (P2WPKH); and
* Pay-to-Taproot (P2TR).

Spends from these three addresses can be produced only through signatures from the
private key associated with the `keyBytes`. Each ::Beacon:: is of the type
[SingletonBeacon].

It takes the following inputs:

* `identifier` - A valid **did:btc1** identifier
* `keyBytes` - a compressed SEC encoded secp256k1 public key
* `network` - the bitcoin network used for the identifier

It returns the following output:

> * `services` - an array of beacon services objects containing the following properties

Reviewer: ```suggestion
* `services` - an array of BTC1 Beacon services objects containing the following properties
```

The Deterministically Generate Beacon Services algorithm generates three
::Beacons:: from the single `keyBytes` value used to generate the deterministic
**did:btc1**, one for each of the following three Bitcoin address types for the
Bitcoin `network` specified by the DID:

* Pay-to-Public-Key-Hash (P2PKH)
* Pay-to-Witness-Public-Key-Hash (P2WPKH); and
* Pay-to-Taproot (P2TR).

Spends from these three addresses can be produced only through signatures from the
private key associated with the `keyBytes`. Each ::Beacon:: is of the type

> [SingletonBeacon].

Reviewer: I wonder if this should be Each BTC1 Beacon is a Singleton Beacon. With the terms referenced. This goes to the singleton beacon section does it?

Author: Sounds good. Yeah, it should have :: around it. Will fix.

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier; REQUIRED; string
* `keyBytes` - a compressed SEC encoded secp256k1 public key; REQUIRED; bytes
* `network` - the Bitcoin network used for the identifier; REQUIRED; string; default=`"bitcoin"`

It returns the following output:

* `services` - an array of ::BTC1 Beacon:: service objects containing the following properties
   * `type` - described the kind of service being defined used to determine how to produce updates; 
      REQUIRED; string; default=`"SingletonBeacon"`
   * `id` - the **did:btc1** identifier controlling the beacon including the DID fragment pointing
      to the location of a key in the DID document; REQUIRED; string

>    * `serviceEndpoint` - a P2PKH, P2WPKH or P2TR bitcoin address where beacon signals can be broadcast; REQUIRED; string

Reviewer: I think this is fine in this PR, but just want to flag that we will need to update this later regarding the new, still slightly debated, type structure for Beacons. i.e. type: ["BTC1Beacon", "SingletonBeacon"]. Or the other variation

part of the resolution request.

It takes the following inputs:

* `identifier` - A valid **did:btc1** identifier
* `identifierComponents` - An object containing the following properties
  * `idType` - the type of identifier (KEY or EXTERNAL)
  * `version` - the identifier version

>   * `network` - the identifier bitcoin network

Reviewer: ```suggestion
  * `network` - the identifier Bitcoin network
```

Could you just check the whole PR for this catch @jintekc

  * `genesisBytes` - the originating public key bytes

> * `resolutionOptions` - A set of optional parameters used for resolution

Reviewer: Should use style I referred to above I feel

##### External Resolution

The External Resolution algorithm externally retrieves an ::intermediate DID document::
as an `intermediateDocumentRepresentation`, either by retrieving it from

> ::ContentiAddressable Storage:: (CAS) or from the ::Sidecar Data:: provided as

Reviewer: ```suggestion
::Content Addressable Storage:: (CAS) or from the ::Sidecar Data:: provided as
```

resolved. 

It takes the following inputs:

* `initialDocument` - the ::initial DID document:: that was used to initiate the **did:btc1**
   identifier being resolved as verified by the [Resolve Initial DID Document] algorithm.
   A DID Core conformant DID document; REQUIRED; object
* `resolutionOptions` - options that extends the default options per
   the [DID Resolution specification](https://w3c.github.io/did-resolution/#did-resolution-options);

>    OPTIONAL; object

Reviewer: This this is okay. Just still wondering how to indicate that this set of options below is not intended to be exhaustive

Author: I'll add a note

```

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier; REQUIRED; string
* `resolutionOptions` - an object that extends the default options per
   the [DID Resolution specification](https://w3c.github.io/did-resolution/#did-resolution-options);
   the below list is not intended to be exhaustive; OPTIONAL; object
   * `versionId` - the version of the DID and DID document, an incrementing integer starting
      from 1; OPTIONAL; integer
   * `versionTime` - a timestamp used during resolution as a bound for when to
      stop resolving; OPTIONAL; integer
   * `sidecarData` - data necessary for resolving a DID such as ::BTC1 Updates:: and
      ::SMT proofs::; OPTIONAL; object.
   * `network` - the Bitcoin network used for resolution; OPTIONAL; string; default=`"bitcoin"`
```

point the current `contemporaryDIDDocument` is returned.

It takes the following inputs:

* `contemporaryDIDDocument` - the DID document for the **did:btc1** identifier 
   being resolved that is current at the blockheight of the `contemporaryBlockheight`. 
   A DID Core conformant DID document; REQUIRED; object
* `contemporaryBlockheight` - a Bitcoin blockheight identifying the contemporary time 
   at which the resolution algorithm has reached as it traverses each block in the 
   blockchain history. an integer greater of equal to 0; REQUIRED; integer
* `currentVersionId` - the version of the ::contemporary DID document::. An integer starting from 

>    1 and incrementing by 1 with each ::DID Update Payload:: applied to the DID document; REQUIRED; integer

Reviewer: payload to update pls

* `targetTime` - a 64-bit UNIX timestamp that can be used to target specific historical states of a DID document.
   Only ::Beacon Signals:: included in the Bitcoin blockchain before the `targetTime` are processed by the
   resolution algorithm; REQUIRED; integer
* `didDocumentHistory` - an ordered array of DID documents from version 1 up to the
   current version; REQUIRED; array
* `updateHashHistory` - an ordered array of SHA256 hashes of ::DID Update Payloads:: 
   that have been applied to the DID document by the resolution algorithm in order 
   to construct the `contemporaryDIDDocument`; REQUIRED; array
* `signalsMetadata` - a Map from Bitcoin transaction identifiers of ::Beacon Signals:: 
   to a struct containing ::Sidecar Data:: for that signal provided as part of the
   `resolutionOptions`; REQUIRED; map containing the following properties:
   * `updatePayload` - a ::DID Update Payload:: which SHOULD match the update announced
      by the ::Beacon Signal::. In the case of a ::SMT:: proof of non-inclusion no
      DID Update Payload MAY be provided; OPTIONAL; object
   * `proofs` - a ::Sparse Merkle Tree:: proof that the provided `updatePayload` value is

>       the value at the leaf indexed by the **did:btc1** being resolved. REQUIRED; TODO: What exactly this

Reviewer: @KDean-Dolphin can you chime in here. This is defined somewhere now (in your pending PR.)  So happy to get this in as is. But we need to track that this TODO needs fixing asap

      structure is needs to be defined.
* `network` -  a string identifying the Bitcoin network of the **did:btc1** identifier.
   The algorithm MUST query the Bitcoin blockchain identified by the `network`; REQUIRED; string

It returns the following output once either `targetTime` or `targetVersionId` have been reached:

* `contemporaryDIDDocument` - A ::contemporary DID document:: conformant to DID Core resolved using updates found
   on-chain from genesis to `contemporaryBlockheight`.

The steps are as follows:

1. Set `contemporaryHash` to the result of passing `contemporaryDIDDocument` into the 

>    [JSON Canonicalization and Hash] algorithm.

Reviewer: ```suggestion
   [JSON Canonicalization and Hash] algorithm.
```

Wondering how we handle Ryans desire to move to RDFC. I am inclined to put a flag at the front of the spec. But leave the spec as is for now.

I do not think we are going to be able to get our libraries in sync and aligned using RDFC in time. Wheras today, they are in sync and working. I do not see a security concern with this personally. It is clearly a distinct crypto suite. And these are just going on the test network anyway. We are not version 1 yet. This is one of the areas we expect to change

   property.
1. For `update` in `orderedUpdates`:
    1. If `update.targetVersionId` is less than or equal to `currentVersionId`,
       push `contemporaryHash` into the `updateHashHistory` list and run the

>        [Confirm Duplicate Update] algorithm passing in `update` and `updateHashHistory`.

Author: I think it makes it overly explicit, which is our goal, but idk that it adds much value since its just var naming. Happy to make the change throughout though to remain consistent.

>        [Confirm Duplicate Update] algorithm passing in `update` and `updateHashHistory`.

Reviewer: Fair enough. Well I definitely want to change anywhere we use `updatePayload`. That should either be `update` or `btc1Update`. I feel for 4 chars we get more clarity which is worthwhile and it maps to our terms

##### Process Beacon Signals

> The Process Beacon Signals algorithm processes each ::Beacon Signal:: by attempting to retrieve and
> validate an announce ::DID Update Payload:: for that signal according to the type of the ::Beacon::.

Reviewer: ```suggestion
The Process Beacon Signals algorithm processes each ::Beacon Signal:: by attempting to retrieve a ::BTC1 Beacon Announcement:: for a specific **did:btc1** identifier.
Then the algorithm retrieves and validates the ::BTC1 Update:: committed to by the announcement. 
The ::Beacon Type:: of the ::BTC1 Beacon:: that broadcast the ::Beacon Signal:: defines the algorithms to retrieve and validate 
::BTC1 Beacon Announcements:: and their associated ::BTC1 Updates"" against the ::Beacon Signal::.
```

It takes the following inputs:

* `beaconSignals` - An array of struct representing ::Beacon Signals:: retrieved through executing
   the [Find Next Signals] algorithm; REQUIRED; object containing the follow properties:
   * `beaconId` - The id for the ::Beacon:: that the `signal` was announced by; REQUIRED; string.
   * `beaconType` - The type of the ::Beacon:: that announced the `signal`; REQUIRED; string.
   * `tx` - The Bitcoin transaction that is the ::Beacon Signal::; REQUIRED; string.
* `signalsMetadata` - A Map from Bitcoin transaction identifiers of ::Beacon Signals:: 
   to a struct containing ::Sidecar Data:: for that signal provided as part of the
   `resolutionOptions`; REQUIRED; object containing the following properties:
   * `updatePayload` - A ::DID Update Payload:: which SHOULD match the update announced
      by the ::Beacon Signal::. In the case of a ::SMT:: proof of non-inclusion no
      DID Update Payload MAY be provided; OPTIONAL; object.
   * `proofs` - A ::Sparse Merkle Tree:: proof that the provided `updatePayload` value is
      the value at the leaf indexed by the **did:btc1** being resolved; OPTIONAL; object.

It returns the following output:

> * `updates` - an array of ::DID Update Payloads::

Reviewer: ```suggestion
* `btc1Updates` - an array of ::BTC1 Updates::
```

   ```{.json include="json/CRUD-Operations/Update-apply-did-update.json"}
   ```
1. Set `expectedProofPurpose` to `capabilityInvocation`.

> 1. Set `mediaType` to `application/ld+json`?

Author: oops! good catch :smile: thanks @parasyte

following the [BIP340 Data Integrity Cryptosuite](https://dcdpr.github.io/data-integrity-schnorr-secp256k1/#instantiate-cryptosuite)
with a `proofPurpose` of `capabilityInvocation`.

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier; REQUIRED; string.
* `sourceDocument` - the DID document being updated; REQUIRED; object.
* `sourceVersionId` - the version of the DID and DID document, an incrementing integer
   starting from 1; OPTIONAL; integer.
* `documentPatch` - JSON Patch object containing a set of transformations to
   be applied to the `sourceDocument`. The result of these transformations MUST
   produce a DID document conformant to the DID Core specification; REQUIRED; object.
* `verificationMethodId` - identifier for a `verificationMethod` within the
   `sourceDocument`. The `verificationMethod` identified MUST be a
   [BIP340 Multikey](https://dcdpr.github.io/data-integrity-schnorr-secp256k1/#multikey);
   REQUIRED; string.
* `beaconIds` - an array containing the IDs that correspond to ::Beacons:: in the DID
   document. These MUST identify service endpoints with one of the three ::Beacon Types::

>    `SingletonBeacon`, `CIDAggregateBeacon`, and `SMTAggregateBeacon`; REQUIRED; array.

Reviewer: ```suggestion
   `SingletonBeacon`, `MapBeacon`, and `SMTBeacon`; REQUIRED; array.
```

   `invalidDidUpdate` error.
1. Set `sourceHashBytes` to the result of passing `sourceDocument` into
   the [JSON Canonicalization and Hash] algorithm.
1. Set `didUpdatePayload.sourceHash` to the base64 of `sourceHashBytes`.
1. Set `targetHashBytes` to the result of passing `targetDocument` into
   the [JSON Canonicalization and Hash] algorithm.
1. Set `didUpdatePayload.targetHash` to the base64 of `targetHashBytes`.
1. Set `didUpdatePayload.targetVersionId` to `sourceVersionId + 1`
1. Return `didUpdatePayload`.

#### Invoke DID Update Payload

The Invoke DID Update Payload algorithm etrieves the `privateKeyBytes` for the
`verificationMethod` and adds a capability invocation in the form of a Data
Integrity proof following the [ZCAP-LD](https://w3c-ccg.github.io/zcap-spec/)
and Verifiable Credentials (VC) Data Integrity specifications.

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier; REQUIRED; string.

> * `didUpdatePayload` - an unsigned ::DID Update Payload::; REQUIRED; object.

Reviewer: payload to btc1Update

the ::Beacon::.

It takes the following inputs:

* `identifier` - a valid **did:btc1** identifier; REQUIRED; string.
* `sourceDocument` - the DID document being updated; REQUIRED; object.
* `beaconIds` - an array containing the IDs that correspond to ::Beacons:: in the DID
   document. These MUST identify service endpoints with one of the three ::Beacon Types::
   `SingletonBeacon`, `CIDAggregateBeacon`, and `SMTAggregateBeacon`; REQUIRED; array.

> * `didUpdateInvocation` - a ::DID Update Payload:: object invoked and secured with

Reviewer: This would require a pass to make sure we are referring to the right things in the right places

> * `didUpdateInvocation` - a ::DID Update Payload:: object invoked and secured with

Author: Sounds good

   signature / proof data; REQUIRED; object.

It returns the following output:

* `signalsMetadata` - a mapping from Bitcoin transaction identifiers of ::Beacon Signals:: 
   to a struct containing ::Sidecar Data:: for that signal, provided as part of the
   `resolutionOptions`; a map with the following properties:

>    * `updatePayload` - a ::DID Update Payload:: which SHOULD match the update announced

Reviewer: payload to btc1Update
